cpp/src/arrow/compute/kernels/vector_nested.cc

```diff
@@ -157,6 +157,43 @@ class ListParentIndicesFunction : public MetaFunction {
   }
 };
 
+const FunctionDoc list_element_doc(
+    "Compute elements using of nested list values using an index",
+    ("`lists` must have a list-like type.\n"
+     "For each value in each list of `lists`, the element at `index`\n"
+     "is emitted."),
+    {"lists", "index"});
+
+template <typename Type, typename IndexType>
+Status ListElement(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+  using ListArrayType = typename TypeTraits<Type>::ArrayType;
+  using IndexScalarType = typename TypeTraits<IndexType>::ScalarType;
+  ListArrayType list_array(batch[0].array());
+  const auto& index_scalar = batch[1].scalar_as<IndexScalarType>();
+  auto index = index_scalar.value;
+  ScalarVector scalars;
```

Comment:
nit: you could `scalars.reserve(list_array.length())`

Reply:
Yes, is a good idea, at that moment I was not clear that the final output will have of the same length.